[GeoMechanicsApplication] Extract computation of integration coefficients

avdg81 opened the pull request:
**📝 Description**
Extracted the computation of a vector of integration points from the integration point loops in (virtually) all implementations of `CalculateAll`. Also modified a few function parameters types from reference to reference-to-const. In that way, it becomes clearer that the supplied object won't be modified by the callee.

avdg81 committed:
Use a local variable to store the permeability matrix

avdg81 committed:
When computing the permeability matrix, don't modify the element variables

avdg81 committed:
Extracted the computation of the integration coefficients

avdg81 committed:
When computing the integration coefficient, only pass the integration point at hand

avdg81 committed:
Replaced a raw for loop by an STL algorithm

avdg81 committed:
Factored out the computation of the integration coefficients

avdg81 committed:
Reverted some unintended changes

rfaasse commented on applications/GeoMechanicsApplication/custom_elements/U_Pw_small_strain_element.cpp:
```diff
@@ -1287,16 +1292,16 @@ void UPwSmallStrainElement<TDim, TNumNodes>::CalculateAndAddCompressibilityMatri
     KRATOS_TRY
 
-    rVariables.PPMatrix = GeoTransportEquationUtilities::CalculatePermeabilityMatrix<TDim, TNumNodes>(
+    const auto permeability_matrix = GeoTransportEquationUtilities::CalculatePermeabilityMatrix<TDim, TNumNodes>(
```
Just double checking, this PPMatrix is never used somewhere without assigning it first right?

avdg81 replied:
I have double-checked that whenever `PPMatrix` is being used, we first assign to it and only then read from it. So unless I'm overlooking something here, the above change should not have any impact elsewhere in the code.

rfaasse reviewed (commented):
Nice PR, quite some changes, but a very precise scope! I have one question left, I believe we discussed it already, but just double checking before approving

markelov208 reviewed (commented):
Hi Anne, nice work: a single point PR, a step forward to change from the integration point level to the element level and to get rid of ElementVariables. I have only one question: why not to create a static function to calculate integration_coefficients?

WPK4FEM commented on applications/GeoMechanicsApplication/custom_elements/U_Pw_small_strain_FIC_element.cpp:
```diff
@@ -468,12 +475,10 @@ void UPwSmallStrainFICElement<TDim, TNumNodes>::CalculateAll(MatrixType& rLeftHa
         this->InitializeBiotCoefficients(Variables, hasBiotCoefficient);
 
-        // Compute weighting coefficient for integration
-        Variables.IntegrationCoefficient =
-            this->CalculateIntegrationCoefficient(IntegrationPoints, GPoint, Variables.detJ);
+        Variables.IntegrationCoefficient = integration_coefficients[GPoint];
 
         Variables.IntegrationCoefficientInitialConfiguration = this->CalculateIntegrationCoefficient(
-            IntegrationPoints, GPoint, Variables.detJInitialConfiguration);
```
This piece of code seems to use the same function with the same input. Therefore I think that detJ == detJInitialConfiguration. In the new implementation storing the list of detJ for the integrationpoints would then avoid the second call to CalculateIntegrationCoefficient.

avdg81 replied:
I'm not sure. `ElementVariables::detJ` is taken from `ElementVariables::detJContainer`, which is computed by calling member `InitializeElementVariables`. `ElementVariables::detJInitialConfiguration` on the other hand is computed by calling member `CalculateKinematics` (which in turn calls member `CalculateDerivativesOnInitialConfiguration`). So they are probably two different things. However, it is unclear to me where `ElementVariables::detJInitialConfiguration` is being used and whether that is correct. To be investigated.

WPK4FEM reviewed (dismissed):
The number of times you had to do the same action reveals how many copies exist within our code base. Your change looks to be consequently executed for all of them.
Remarks that I would like to make would without exception refer to the existing situation that you start improving here, so they would not be productive here.

Please go ahead and merge this, such that a next step can be taken.

avdg81 committed:
Two member functions are no longer `virtual` and they are `const` now

Calculation of the integration coefficient does not require the member
function to be non-`const`. Also, these member functions were declared
`virtual`, but never overridden.

avdg81 committed:
Extracted member functions to calculate the integration coefficients

To reduce the amount of duplicated code. As a result, the variables that
store the result could be made `const`.

avdg81 committed:
Simplified two types and reformatted some code

rfaasse reviewed (approved):
Nicely reduced the duplications, looks good to go to me!

WPK4FEM reviewed (approved):
Looks consistent to me. No comments from my side.
